### ISSUE TITLE
A* Shortest paths with observations

### ISSUE THREAD

**Issue by Zannick** [Zannick/logic-graph#190]
Problem: Our shortest paths table includes the minimum possible distance over any edge in the graph, including edges we don't have access to.

For example, AV2 uses a Map area with spots for each room with a fast-travel destination, and a singular exit to that room. If the fastest route from a starting spot S to a goal spot G goes through a fast-travel destination D that we don't have direct access, the A* search will prioritize going to the Map spot that links to D, but be unable to use the edge. As a result, A* will attempt to route around the "obstacle", which would include generating one state for each Map spot to put into the queue. One of those that we can fast-travel to will be the closest without the fast-travel. (In a few cases, the fastest route will involve fast-travel to move somewhere to invoke a different warp.) But from that fast-travel point that we can use, the fastest possible route may be to fast-travel to D--this might be true for all fast-travel spots, in which case they all have the same priority in the queue.

When we make a change to state (such as entering a new area and updating a context variable), then the fast-travel warp creates a distinct state from the other ones we'd already seen. If we have multiple cheap steps we can make, with the shortest paths pointing to use fast-travel like this, we may end up with many times the number of Map spots in the queue, none of which can actually make it faster than picking a different fast-travel point and moving from there. With a limited queue size (and/or limited number of states to process), this can prevent a route from being found.

Despite those differences in uniqueness, the failure to take certain edges is observable. If we observe why we can't take this edge in our first pass, i.e. we haven't visited the fast-travel point before, can we keep that observation to say, "we can't use fast-travel here unless we pass one of these observations now"? Then, when we make a subtle context change before starting fast-travel, we can say, that change doesn't affect anything, and discard the option to perform fast-travel, since it's unlikely to produce a better option.

It may perhaps be even better to have observations on entire routes between S and G, and cache or precalculate them. Our shortest path only includes edges (and thus not actions), but an A* search is guaranteed to find the shortest route with actions if one exists (notwithstanding any shortest routes that take more actions than our max actions)--observing this route means that any other A* search between S and G that matches this observation can take at least this route. If we include observations about *shorter* routes, such as the fast-travel way, then we can check observations in order. If we include negative observations, we may be able to prove that any state matching the observation has the given route as its shortest, but the hit rate may decrease, which over time would result in more entries in the cache.

**Comment by Zannick** [comment]
The proposal for a data structure for routes  looks like this:
- A map from S to G to an observation trie.
- The observation trie is a matcher trie but with a default value rather than a root observation like position. The default is either the shortest free route, or None if no such route exists.

As long as we trust the fastest routes provided are the actual fastest routes, then we shouldn't need to check whether a set of observations being inserted is a superset of a shorter route (since it would be possible to take that route if so), nor even if it beats the default.

We would use the route from the observation trie as our maximum time for a route, but we can only decide a route is the shortest possible by checking the shortest paths table&mdash;in all other cases we would still perform a search, but with a much lower maximum time. In fact, we could check whether the default (the shortest possible free route) is the shortest possible route in general, and skip having to deal with an observation matcher trie for those entirely.

This structure may get very large, so we should consider how we might include it in the db. 
* Key: Encoded `<S><G><observations>`
* Value: Encoded route

Since the route for the given observations should be unique, we don't have to worry too much about write conflicts. Read-write conflicts are then resolved by reads simply being stale (missing a route), which would lead to a duplicate write. A reader could either read and test single observations sequentially from the db (multiple trips), or read the entire trie data from the db before testing.

**Comment by Zannick** [comment]
Finally looking at building this. It might be worth building a shortest paths table using only free movement. It might even be possible to generate the whole observation table immediately based on finding all the paths up to the default free path time, except that not all pairs of Spots will have a default free path between them. (Actually, we can't generate observations without having a state beforehand--to do this ahead of time would require a rules analyzer that can generate conditional observations, or grouping the edges together like condensed edges, which would probably result in a lot of duplication of requirements.)

Building observations with negative assertions will be strange. In practice we have to find the failing edge in any faster route. If we generate all the routes then we can easily do this... but if we generated all the routes we would just try the observations like normal. Without that, as we do our A* from our initial state to find new routes, edges that can't be traversed may need to have their observations (or at least the one component that caused the access rule to fail, which might be harder) attached to the parent state and applied to the route we find if we use one of the child edges that worked. Yet we don't really need those negative observations in every case, only in cases where that edge would get us a faster route. This might be more complex than it would be worth.

Here's a proposed algorithm:
1. Use the stored observation trie to find the fastest known route available. If it's the shortest path overall, we're done.
2. Take the time of that fastest known route, or the max time on the current search if there was no such route.
3. A* with that max time to find a faster route.
4. If we find one, add it to the observation trie.

And for using the observation trie to find the fastest known route, either:
* check against the observation lists in order of the route speed
* lookup all the matching results in the matcher trie and taking the smallest or default or none

**Comment by Zannick** [comment]
Also, our observation tries right now require the root node be a partitioned observation (typically the position). We could modify the matchertrie module to support doing without, but we could also structure as a map
* dest -> observation trie (which starts with src position as usual)

**Comment by Zannick** [comment]
This feature would be helpful for cases described in #216

**Comment by Zannick** [comment]
There's another issue here if we want to consider observations as infallible (which would require that we observe penalty conditions...): the main Matcher classes are written to accept multiple values in a HashSet and not overwrite or remove existing ones. It might be better to write Matchers that store a value in an Option.

**Comment by Zannick** [comment]
Potentially we should use a DirectPaths structure that maps the *check* to the partial route, rather than *a spot* to the partial route, just in case we search a shortest path to the spot containing a check but remain unable to access it.
